File: BasicNet/MultiThreadServer.cs
using System;
using GaiaNet.GaiaNets;
using System.Net.Sockets;
using System.IO;
using System.Net;
using System.Threading;
using GaiaNet.Command;
using System.Text;
using GaiaNet.Relay;
using GaiaNet.FilesTransfer;
using GaiaNet.HolePunching;
using System.Collections.Generic;

namespace GaiaNet.BasicNet
{
    public class MultiThreadServer
    {
        public Net gaiaNet = null;
        private int _port;
        public String BasePath { get; set; }
        private TcpListener _server;
        private Boolean _isRunning;
        private Dictionary<string, Socket> reverseProxys = new Dictionary<string, Socket>();
        private static readonly log4net.ILog log = log4net.LogManager.GetLogger(
            System.Reflection.MethodBase.GetCurrentMethod().ReflectedType);

        public MultiThreadServer(int port):this(port, null) {
            _port = port;
        }

        public MultiThreadServer(int port, Net net) {
            _port = port;
            try{
                _server = new TcpListener(IPAddress.Any, _port);
                BasePath = Directory.GetCurrentDirectory();
                _isRunning = true;
            }
            catch (IOException e){ log.Error(e); }
            gaiaNet = net;
        }

        public void serverRun(){
            _server.Start();
            log.Info(String.Format("MultiThreadServer has started at: {0}", _port));
            
            while (_isRunning){
                try {
                    Socket newClient = _server.AcceptSocket();
                    NetType type = new NetHeader().ReadHeader(newClient);
                    log.Info(String.Format("Connection: {0} with GaiaType {1}",newClient.RemoteEndPoint,type.ToString()));

                    switch (type)
                    {
                        case NetType.Command:
                            // new Thread(()=> new CommandHandler().handle(newClient)).Start();
                            break;
                        case NetType.File:
                            new Thread( ()=>new RecvFiles(newClient).Recv() ).Start();
                            break;
                        case NetType.GaiaNet:
                            //new Thread(()-> this.gaiaNet.net.handle(socket, logServer)).start();
                            break;
                        case NetType.Relay:
                            new Thread( ()=>new TcpRelay(newClient).Relay() ).Start();
                            break;
                        case NetType.HolePunch:
                            // new Thread( ()=>new TcpPunchServer().handle(newClient) ).Start();
                            break;
                        case NetType.ReverseProxy:
                            // new Thread( ()=>new ReverseProxy(reverseProxys).handle(newClient) ).Start();
                            break;
                        default:
                            log.Info("MultiThreadServer: Unknown Socket.");
                            break;
                    }

                } catch (IOException e) {
                    log.Error(e);
                }

            }
        }

        public void stop() {
            try {
                _server.Stop();
            } catch (IOException e) { log.Error(e); }
        }

        // Just print out the stream with hex string
        //public void printStream(Socket socket, DataInputStream dis){
        //    try {
        //        byte[] byt = new byte[10];
        //        dis.read(byt);
        //        System.out.print(ByteTools.HexString(byt));
        //    } catch (IOException e) {
        //        e.printStackTrace();
        //    }

        //}
    }
}

File: old java code/src/GaiaNet/NetTransfer/MultiThreadServer.java
package GaiaNet.NetTransfer;

import GaiaNet.Common.ByteTools;
import GaiaNet.GaiaNet.Net;

import java.io.*;
import java.net.ServerSocket;
import java.net.Socket;
import java.nio.file.Path;
import java.util.Hashtable;
import java.util.concurrent.locks.ReadWriteLock;
import java.util.concurrent.locks.ReentrantReadWriteLock;
import java.util.logging.FileHandler;
import java.util.logging.Logger;
import java.util.logging.SimpleFormatter;

public class MultiThreadServer {
    private static final Logger logServer = Logger.getLogger( MultiThreadServer.class.getName() );
    public FileHandler fileHandler;  // For log file.
    public Net gaiaNet = null;
    private int port;
    private String basePath;
    ServerSocket sSocket;
    Hashtable<Integer, ReadWriteLock>  rwLocks = new Hashtable<>();

    public String getBasePath() {
        return basePath;
    }
    public void setBasePath(String basePath) {
        this.basePath = basePath;
    }

    public MultiThreadServer(int port) {
        this.port = port;
        try {
            sSocket = new ServerSocket(port);
            basePath = System.getProperty("user.dir");
            fileHandler = new FileHandler("./log/MultiThreadServer.log");   // For log file.
            fileHandler.setFormatter(new SimpleFormatter());
            logServer.addHandler(fileHandler);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public MultiThreadServer(int port, Net net) {
        this(port);
        this.gaiaNet = net;
    }

//
    public void serverRun(){
        int flag;
        while (true){
            try {
                Socket socket = sSocket.accept();
                DataInputStream dis = new DataInputStream(socket.getInputStream());
                flag = dis.readInt();
                GaiaType type = GaiaType.values()[flag];

                if (type == GaiaType.Command){
                    CommandHandler cmdHandler = new CommandHandler();
                    new Thread(()-> cmdHandler.handle(socket, logServer)).start();
                } else if (type == GaiaType.File){  // file transfer
                    FileHeader fileHeader = new FileHeader();
                    fileHeader.receive(dis);
                    logServer.info("File from: "+socket.getRemoteSocketAddress()+ " File Name: "+fileHeader.name+" Type: "+fileHeader.type);
                    if (fileHeader.type == 0)  createFile(socket,dis,fileHeader);
                    if (fileHeader.type == 1)  new Thread(()-> threadRcvFile(socket,dis,fileHeader)).start();
                } else if (type == GaiaType.GaiaNet){
                    new Thread(()-> this.gaiaNet.net.handle(socket, logServer)).start();
                }
                else if (flag==100){
                    printStream(socket, dis);
                } else {
                    System.out.println("MultiThreadServer: Unknown Socket.");
                }
            } catch (IOException e) {
                e.printStackTrace();
            }

        }
    }

//  used to receive files. first read the [name,star,end], then write in file.
    private void threadRcvFile(Socket socket, DataInputStream dis, FileHeader fileHeader) {
        int len;
        try {
            String fileName = fileHeader.name;
            Path filePath = Path.of(basePath, fileName);
            File file = filePath.toFile();
            if (!file.exists()){
                throw new IOException("File is not created!!");
            }

            long start = fileHeader.start;
            long end = fileHeader.end;
            long index = fileHeader.index;
            System.out.printf("thread: "+Thread.currentThread().getName()+" is ready to receive: "+fileName);
            System.out.printf(", start: %d, end: %d \n", start,end);
            RandomAccessFile rw = new RandomAccessFile(file, "rw");
            byte[] byt = new byte[500 * 1024];
            while ((len = dis.read(byt,0, 500 * 1024))!=-1){
                rwLocks.get(file.toString().hashCode()).writeLock().lock();
                rw.seek(start+index);
                rw.write(byt,0, len);
                index += len;
                System.out.printf("receiving: "+Thread.currentThread().getName()+", start: %d, end: %d, index: %d\n", start,end,index);
                rwLocks.get(file.toString().hashCode()).writeLock().unlock();
            };
            rw.close();

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

//  used to write the
    private void createFile(Socket socket, DataInputStream dis, FileHeader fileHeader){
        String fileName;
        try {
            fileName = fileHeader.name;
            long size = fileHeader.size;
            File f = Path.of(basePath, fileName).toFile();
            if ( f.exists() && (f.length()==size) ){
                //  read configuration
            }else {
                RandomAccessFile rw = new RandomAccessFile(f, "rw");
                rw.setLength(size);
            }
            rwLocks.put(f.toString().hashCode(), new ReentrantReadWriteLock());
            System.out.println(Thread.currentThread().getName()+", New file: " +fileName+", size: "+size);
        } catch (IOException e) {
            e.printStackTrace();
        }

    }

    public void stop() {
        try {
            sSocket.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    // Just print out the stream with hex string
    public void printStream(Socket socket, DataInputStream dis){
        try {
            byte[] byt = new byte[10];
            dis.read(byt);
            System.out.print(ByteTools.HexString(byt));
        } catch (IOException e) {
            e.printStackTrace();
        }

    }

}
